contents/dataframes_performance.md

```diff
@@ -4,8 +4,237 @@ So far, we didn't think about making our `DataFrames.jl` code **fast**.
+```jl
+s = """
+    # allocation # hide
+    df = responses()
+    @allocated select(df, :id, :q1)
+    """
+sco(s; process=string, post=plainblock)
+```
+
+```jl
+s = """
+    # non allocation # hide
+    df = responses()
+    @allocated select!(df, :id, :q1)
+    """
+sco(s; process=string, post=plainblock)
+```
+
+As we can see `select!` makes less allocations that `select` so it should be faster while consuming less memory.
+
+### Copying vs Not Copying Columns
+
+There are **two ways to access a DataFrame column**.
+They differ in how they are accessed: one creates a "view" to the column without copying and the other creates a whole new column by copying the original column.
+
+The first is the regular dot `.` operator followed by the column name, like in `df.Col`.
+This kind of access **does not copy** the column `Col`.
+Instead `df.Col` creates a "view" which is a link to the original column without performing any allocation.
+Additionally, the syntax `df.Col` is the same as `df[!, :Col]` with the bang `!` as the row selector.
+
+The second way to access a `DataFrame` column is the `df[:, :Col]` with the colon `:` as the row selector.
+This kind of access **does copy** the column `Col`, so beware that it may produce unwanted allocations.
+
+As before, let's try out our two ways to access a column in the responses `DataFrame`:
+
+```jl
+s = """
+    # allocation # hide
+    df = responses()
+    @allocated col = df[:, :id]
+    """
+sco(s; process=string, post=plainblock)
 ```
+
+
+```jl
+s = """
+    # non allocation # hide
+    df = responses()
+    @allocated col = df[!, :id]
+    """
+sco(s; process=string, post=plainblock)
+```
+
+When we access a column without copying it we are making zero allocations and our code should be faster.
+So, if you don't need a copy always access your `DataFrame`s columns with `df.col` or `df[!, :col]` instead of `df[:, :col]`.
+
+### CSV.read versus CSV.File
+
+If you take a look at the help output for `CSV.read` you will see that it is a convenience function identical to a function called `CSV.File` with the same keywords arguments.
+Both `CSV.read` and `CSV.File` will read the contents of a CSV file, but they differ in the default behavior.
+**`CSV.read`, by default, will not make copies** of the incoming data.
+Instead, `CSV.read` will pass all the data to the second argument (known as "sink").
+
+So, something like this:
+
+```julia
+df = CSV.read("file.csv", DataFrame)
+```
+
+will pass all the incoming data from `file.csv` to the `DataFrame` sink, thus returning a `DataFrame` type that we store in the `df` variable.
+
+For the case of **`CSV.File` the default behavior is the opposite: it will make copies of every column contained in the CSV file**.
+Also the syntax is slight different, we need to wrap anything that `CSV.File` returns in a `DataFrame` constructor function:
+
+```julia
+df = DataFrame(CSV.File("file.csv"))
+```
+
+Or, in a more preferred idiomatic syntax with the pipe `|>` operator:
```

Comment:
I remember a discussion a while back where Tamas Papp said that almost in all cases
```
a = f(x)
b = g(a)
```
is better than
```
a = g(f(x))
```
or
```
a = f(x) |> g
```
The reason is pretty straightforward: if you want to debug the first way, you can easily pin the debugger to a specifc point in the operation. Also, you can easily comment out or remove a part of the operation and it's more clear because the intermediate variable can get a clear name.

That said, I don't think it's wrong per se and am curious for your thoughts. I love the pipe operator when I'm doing stuff in the REPL. If you prefer the pipe operator here, then that's fine.

Reply:
If I do JuliaFormatter.jl to the file it fixes all the pipe stuff. This:

`f(x) |> g`

Becomes:

`g(f(x))`

Reply:
JuliaFormatter isn't perfect either. It also has some non-idiomatic defaults

Reply:
I use JuliaFormatter blue code style.
https://domluna.github.io/JuliaFormatter.jl/dev/blue_style/